File: src/lib/chunk_generator.cpp
#include <cmath>

#include <noise/noise.h>

#include "chunk_generator.h"

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wunused-parameter"
std::array<WorldBlock, CHUNK_SIZE_CUBED> ChunkGenerator::perlin(glm::vec3 position) {
  std::array<WorldBlock, CHUNK_SIZE_CUBED> blocks {};
  if (position.y <= 1) {
    for (unsigned int x = 0; x < CHUNK_SIZE; x++) {
      for (unsigned int z = 0; z < CHUNK_SIZE; z++) {
        for (unsigned int y = 0; y < CHUNK_SIZE; y++) {
          unsigned int index { z * CHUNK_SIZE_SQUARED + y * CHUNK_SIZE + x };
          blocks[index] = Block::createWorldBlock("stone");
        }
      }
    }
  } else if (position.y == 2) {
    noise::module::Perlin perlinNoise;
    const int min { 0 };
    const int max { CHUNK_SIZE };

    for (unsigned int blockX = 0; blockX < CHUNK_SIZE; ++blockX) {
      for (unsigned int blockZ = 0; blockZ < CHUNK_SIZE; ++blockZ) {
        double noiseX = static_cast<float>(blockX) / CHUNK_SIZE + position.x;
        double noiseZ = static_cast<float>(blockZ) / CHUNK_SIZE + position.z;

        double noise = std::clamp((1 + perlinNoise.GetValue(noiseX, 0.0, noiseZ)) / 2.0, 0.0, 1.0);
        unsigned int height = static_cast<unsigned int>(std::clamp(static_cast<int>(min + noise * (max + 1 - min)), 0, CHUNK_SIZE));
        for (unsigned int y = 0; y < height; y++) {
          unsigned int index { blockZ * CHUNK_SIZE_SQUARED + y * CHUNK_SIZE + blockX };
          if (y <= height - 3) {
            blocks[index] = Block::createWorldBlock("stone");
          } else {
            blocks[index] = Block::createWorldBlock("grass");
          }
        }
      }
    }
  }
  return blocks;
}
#pragma GCC diagnostic pop

File: src/scripts/chunk_generator.rb
class ChunkGenerator
  def self.flat(pos_x, pos_y, pos_z)
    y = 0
    blocks = []
    CHUNK_SIZE.times do |x|
      CHUNK_SIZE.times do |z|
        index = z * CHUNK_SIZE_SQUARED + y * CHUNK_SIZE + x
        blocks[index] = :dirt
      end
    end
    blocks
  end

  def self.flat_half_and_half(pos_x, pos_y, pos_z)
    y = 0
    blocks = []
    CHUNK_SIZE.times do |x|
      CHUNK_SIZE.times do |z|
        index = z * CHUNK_SIZE_SQUARED + y * CHUNK_SIZE + x
        blocks[index] =
          if x < CHUNK_SIZE_HALVED
            :dirt
          else
            :stone
          end
      end
    end
    blocks
  end

  def self.flat_half_and_half_with_square(pos_x, pos_y, pos_z)
    y = 0
    blocks = []
    CHUNK_SIZE.times do |x|
      CHUNK_SIZE.times do |z|
        index = z * CHUNK_SIZE_SQUARED + y * CHUNK_SIZE + x
        blocks[index] =
          if x > CHUNK_SIZE / 4 && x < 3 * CHUNK_SIZE / 4 &&
              z > CHUNK_SIZE / 4 && z < 3 * CHUNK_SIZE / 4

            :bedrock
          elsif x < CHUNK_SIZE_HALVED
            :dirt
          else
            :stone
          end
      end
    end
    blocks
  end

  def self.flat_random(pos_x, pos_y, pos_z)
    y = 0
    blocks = []
    CHUNK_SIZE.times do |x|
      CHUNK_SIZE.times do |z|
        index = z * CHUNK_SIZE_SQUARED + y * CHUNK_SIZE + x
        blocks[index] = if rand < 0.5
          :dirt
        else
          :stone
        end
      end
    end
    blocks
  end

  def self.flat_with_plus(pos_x, pos_y, pos_z)
    return [] if pos_y != 0

    blocks = []
    CHUNK_SIZE.times do |x|
      CHUNK_SIZE.times do |z|
        CHUNK_SIZE_QUARTERED.times do |y|
          index = z * CHUNK_SIZE_SQUARED + y * CHUNK_SIZE + x
          blocks[index] =
            if y.zero?
              :bedrock
            elsif x == CHUNK_SIZE_HALVED || z == CHUNK_SIZE_HALVED
              :dirt
            end
        end
      end
    end
    blocks
  end

  def self.half(pos_x, pos_y, pos_z)
    blocks = []
    CHUNK_SIZE.times do |x|
      CHUNK_SIZE.times do |z|
        CHUNK_SIZE_HALVED.times do |y|
          index = z * CHUNK_SIZE_SQUARED + y * CHUNK_SIZE + x
          blocks[index] = :dirt
        end
      end
    end
    blocks
  end

  def self.filled(pos_x, pos_y, pos_z, block_id = :dirt)
    blocks = []
    CHUNK_SIZE.times do |x|
      CHUNK_SIZE.times do |z|
        CHUNK_SIZE.times do |y|
          index = z * CHUNK_SIZE_SQUARED + y * CHUNK_SIZE + x
          blocks[index] = block_id
        end
      end
    end
    blocks
  end

  def self.half_sphere(pos_x, pos_y, pos_z)
    blocks = []
    radius = CHUNK_SIZE_HALVED
    CHUNK_SIZE.times do |x|
      CHUNK_SIZE.times do |z|
        8.times do |y|
          block_x = x - CHUNK_SIZE / 2
          block_y = y - CHUNK_SIZE / 2
          block_z = z - CHUNK_SIZE / 2
          if Math.sqrt(block_x * block_x + block_y * block_y + block_z * block_z) < radius
            index = z * CHUNK_SIZE_SQUARED + y * CHUNK_SIZE + x
            blocks[index] = :dirt
          end
        end
      end
    end
    blocks
  end

  def self.perlin(pos_x, pos_y, pos_z)
    blocks = []
    case pos_y
    when 0..1
      blocks = filled pos_x, pos_y, pos_z, :stone
    when 2
      min_height = 0
      max_height = CHUNK_SIZE

      CHUNK_SIZE.times do |block_x|
        CHUNK_SIZE.times do |block_z|
          noise_x = block_x / CHUNK_SIZE + pos_x
          noise_z = block_z / CHUNK_SIZE + pos_z

          generated_noise = (1 + CNoise.get_perlin_noise(noise_x, 0.0, noise_z)) / 2.0
          noise = generated_noise.clamp(0.0, 1.0)
          raw_height = min_height + noise * (max_height + 1 - min_height)
          height = raw_height.clamp(min_height, max_height)

          height.times do |y|
            index = block_z * CHUNK_SIZE_SQUARED + y * CHUNK_SIZE + block_x
            blocks[index] =
              if y <= height - 3
                :stone
              else
                :grass
              end
          end
        end
      end
    end
    blocks
  end
end
